[Feature][Registry] Support etcd as registry

wjf222 opened the pull request:
<!--Thanks very much for contributing to Apache DolphinScheduler. Please review https://dolphinscheduler.apache.org/en-us/community/development/pull-request.html before opening a pull request.-->


## Purpose of the pull request

Close #8975

## Brief change log
Add CRUD
Add Subscribe
Add Lock
Add ConnectStateListener

## Verify this pull request

This change added tests and can be verified as follows:
Add persistTest to test for CRUD
Add lockTest for lock

TODO add connectStateListener Test

wjf222 committed:
Add dependencyes

wjf222 committed:
Add Test and EtcdRegistry Constructor

wjf222 committed:
Add Crud and Pass EtcdRegistryTest

wjf222 committed:
Add acquireLock and releaseLock

wjf222 committed:
Add subscribe

wjf222 committed:
Add ConnectStateListener

wjf222 committed:
fix some bug

wjf222 committed:
Add comment

wjf222 committed:
Add README.md

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistryProperties.java:
```diff
@@ -0,0 +1,31 @@
+    private String authority;
+
+    // RetryPolicy
+    private Long retryDelay=60L;
+    private Long retryMaxDelay=300L;
```
Consider also using `Duration` type, and format the codes please.

wjf222 replied:
The time unit for these two parameters is set by the ChronUnitRetry in jetcd.  The type of these two parameters is long, but the type of retryMaxDuration is Duration.

You can see the reference of these three parameters in EtcdRegistry.java line 51 to 54.

kezhenxu94 replied:
> The time unit for these two parameters is set by the ChronUnitRetry in jetcd. The type of these two parameters is long, but the type of retryMaxDuration is Duration.
> 
> You can see the reference of these three parameters in EtcdRegistry.java line 51 to 54.

Hi, you don't have to just copy the type from JETCD client, using `Duration` as type in our own configurations is for users' convenient, they can just set something like `1s`, `500ms`, without knowing what's the time unit of the config. Also, users don't care what types we pass into `EtcdRegistry.java`, let's provide simplicity to users

wjf222 replied:
> > The time unit for these two parameters is set by the ChronUnitRetry in jetcd. The type of these two parameters is long, but the type of retryMaxDuration is Duration.
> > You can see the reference of these three parameters in EtcdRegistry.java line 51 to 54.
> 
> Hi, you don't have to just copy the type from JETCD client, using `Duration` as type in our own configurations is for users' convenient, they can just set something like `1s`, `500ms`, without knowing what's the time unit of the config. Also, users don't care what types we pass into `EtcdRegistry.java`, let's provide simplicity to users

You are right,I fix this in [Modify the type of delay](https://github.com/apache/dolphinscheduler/pull/10981/commits/ba5f7687f8842427e4fea67b56f7f1125e48b9d4)

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistryProperties.java:
```diff
@@ -0,0 +1,31 @@
+    // RetryPolicy
+    private Long retryDelay=60L;
+    private Long retryMaxDelay=300L;
+    private Duration retryMaxDuration=Duration.ofMillis(1500);;
```
```suggestion
    private Duration retryMaxDuration=Duration.ofMillis(1500);
```

wjf222 replied:
How to undo this operation, I clicked wrongly

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/pom.xml:
```diff
@@ -0,0 +1,43 @@
+            <artifactId>jetcd-core</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>io.etcd</groupId>
+            <artifactId>jetcd-test</artifactId>
+        </dependency>
```
This should be `test` scope

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/pom.xml:
```diff
@@ -0,0 +1,43 @@
+
+    <artifactId>dolphinscheduler-registry-etcd</artifactId>
+
+    <properties>
+        <maven.compiler.source>8</maven.compiler.source>
+        <maven.compiler.target>8</maven.compiler.target>
+    </properties>
```
Don't set this here

wjf222 committed:
Format the pom.xml

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/README.md:
```diff
@@ -0,0 +1,28 @@
+
+# How to use
+
+If you want to set the registry center as mysql,You need to set the registry properties in master/worker/api's appplication.yml
```
```suggestion
If you want to set the registry center as mysql, you need to set the registry properties in master/worker/api's appplication.yml
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/pom.xml:
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
License header

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,301 @@
+    public void put(String key, String value, boolean deleteOnDisconnect) {
+        try{
+            if(deleteOnDisconnect) {
+                // keep the key by lease, if disconnected, the lease will ,the key will delete
```
This sentence is not complete...

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdConnectionStateListener.java:
```diff
@@ -0,0 +1,133 @@
+     * @return current connection channel
+     */
+    private ManagedChannel newChannel(Client client) {
+        try {
+            Field connectField =client.getClass().getDeclaredField("connectManager");
+            if(!connectField.isAccessible()){
+                connectField.setAccessible(true);
+            }
+            Object connection = connectField.get(client);
+            Method channel = connection.getClass().getDeclaredMethod("getChannel");
+            if (!channel.isAccessible()) {
+                channel.setAccessible(true);
+            }
+            return (ManagedChannel) channel.invoke(connection);
+        } catch (Exception e) {
+            throw new RegistryException("Failed to get the etcd client channel", e);
+        }
```
I don't think this is a good way to implement the connectivity state listener, why do you choose to use reflection instead of some native method of jetcd client?

wjf222 replied:
Ok, I will try to use keepalive in lease client to listen for connection status.

wjf222 replied:
> I don't think this is a good way to implement the connectivity state listener, why do you choose to use reflection instead of some native method of jetcd client?

I made the problem complicated. I only need to periodically try to connect to Etcd, and the connection status can be judged by the connection result.
I have removed the reflection related code. Now, i use the client to apply for a lease to determine whether the connection is successful in [Using the lease to listen connection state](https://github.com/apache/dolphinscheduler/pull/10981/commits/d3ecd2a6eed4af52395d2547ef7486cf435d03a1)

kezhenxu94 commented:
Many files are missing license headers

wjf222 committed:
Add license header

caishunfeng commented:
Hi @wjf222 please add the doc of the registry together.

wjf222 committed:
Modify the type of delay

wjf222 commented:
> Many files are missing license headers


I fixed this in the [Add license header](https://github.com/apache/dolphinscheduler/pull/10981/commits/65d0b9d78401478d0f15830b1f949e725771a20a)

wjf222 committed:
Using the lease to listen connection state

wjf222 committed:
[Document] Add readme and comments

wjf222 commented:
> Hi @wjf222 please add the doc of the registry together.

I add the doc in [[Document] Add readme and comments](https://github.com/apache/dolphinscheduler/pull/10981/commits/48dc9982af58b5015110469c3857312de858d0c4).Do you need to add other documents here?

wjf222 committed:
Merge branch 'dev' into wjf-add-etcd-registry

wjf222 committed:
check style

wjf222 committed:
Merge remote-tracking branch 'origin/wjf-add-etcd-registry' into wjf-add-etcd-registry

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/ConnectionListener.java:
```diff
@@ -19,6 +19,10 @@
 
+/**
+ * when the connect state between client and registry center changed,
+ * the onupdate function is triggered
```
```suggestion
 * the {@code onUpdate} function is triggered
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/ConnectionState.java:
```diff
@@ -19,6 +19,9 @@
 package org.apache.dolphinscheduler.registry.api;
 
+/**
+ * Connection State Between client and registry center(Etcd,MySql,Zookeeper)
```
```suggestion
 * Connection state between client and registry center(Etcd, MySql, Zookeeper)
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/Event.java:
```diff
@@ -20,9 +20,13 @@
     private String path;
+    // The value corresponding to the path
     private String data;
+    // The event type {ADD, REMOVE ,UPDATE}
```
```suggestion
    // The event type {ADD, REMOVE, UPDATE}
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/Registry.java:
```diff
@@ -20,27 +20,72 @@
+ * Registry
+ *
+ * <p>
+ * The implementation may throw RegistryException during function call
```
This is verbose as the function signatures already indicate that.

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/Registry.java:
```diff
@@ -20,27 +20,72 @@
+    /**
+     * Watch the change of this path and subpath.
+     * The type of change contains [ADD,DELETE,UPDATE]
+     * @return if there is not a Exception, the result is true.
```
```suggestion
     * @return {@code true} if succeeded.
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/Registry.java:
```diff
@@ -20,27 +20,72 @@
     boolean subscribe(String path, SubscribeListener listener);
 
+    /**
+     * remove the SubscribeListener which subscribe this path
```
```suggestion
     * Remove the path from the subscribe list.
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/Registry.java:
```diff
@@ -20,27 +20,72 @@
     void unsubscribe(String path);
 
+    /**
+     * addd a connection listener to collection
```
```suggestion
     * Add a connection listener to collection.
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/Registry.java:
```diff
@@ -20,27 +20,72 @@
     Collection<String> children(String key);
 
+    /**
+     * @return if key exists,return true
```
```suggestion
     * @return {@code true} if key exists.
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/README.md:
```diff
@@ -0,0 +1,28 @@
+
+# How to use
+
+If you want to set the registry center as mysql,you need to set the registry properties in master/worker/api's appplication.yml
```
```suggestion
If you want to set the registry center as etcd, you need to set the registry properties in master/worker/api's appplication.yml
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdConnectionStateListener.java:
```diff
@@ -0,0 +1,119 @@
+     */
+    private ConnectionState isConnected() {
+        try {
+            // Use Get() to ensure Future completes
```
Remove this. Don't explain codes if they are already very clear.

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdConnectionStateListener.java:
```diff
@@ -0,0 +1,119 @@
+     * @return the current connection state
+     * @throws if there is a exception, return is DISCONNECTED
+     */
+    private ConnectionState isConnected() {
```
```suggestion
    private ConnectionState currentConnectivityState() {
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdConnectionStateListener.java:
```diff
@@ -0,0 +1,119 @@
+            if (currentConnectionState == connectionState) {
+                return;
+            }
+            if (connectionState == ConnectionState.CONNECTED) {
+                if (currentConnectionState == ConnectionState.DISCONNECTED) {
+                    connectionState = ConnectionState.DISCONNECTED;
+                    triggerListener(ConnectionState.DISCONNECTED);
+                }
+            } else if (connectionState == ConnectionState.DISCONNECTED) {
+                if (currentConnectionState == ConnectionState.CONNECTED) {
+                    connectionState = ConnectionState.CONNECTED;
+                    triggerListener(ConnectionState.RECONNECTED);
+                }
+            } else if (connectionState == null) {
+                connectionState = currentConnectionState;
+                triggerListener(connectionState);
+            }
```
Why can't we just simplify this to

```suggestion
            if (connectionState != ConnectionState.SUSPENDED) {
                connectionState = currentConnectionState;
                triggerListener(connectionState);
```

@ruanwenjun I think the api should also notify even if the previous state is suspended and if the listener isn't interested in this they can just ignored.

ruanwenjun replied:
Yes, in fact, for all state changes, we need to notify the listener.
I wrote these logic in MysqlRegistry, since we need to generate `ConnectionState.RECONNECTED` state :), but the state we story in memory is just `RECONNECTED` and `RECONNECTED`.

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdConnectionStateListener.java:
```diff
@@ -0,0 +1,119 @@
+import io.grpc.ManagedChannel;
+
+/**
+ * Get the connection status by listening to the Client's Channel
```
Reword this there is no channel anymore

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdConnectionStateListener.java:
```diff
@@ -0,0 +1,119 @@
+    // A thread pool that periodically obtains connection status
+    private final ScheduledExecutorService scheduledExecutorService;
+    // Client's Channel
+    private AtomicReference<ManagedChannel> channel;
```
This is no needed right?

wjf222 commented:
Thanks for your review, I'll keep commiting after the unit test bug is resolved

wjf222 committed:
Resolve conversation and fix test bug

wjf222 committed:
Unified GRPC version

wjf222 committed:
Merge branch 'dev' into wjf-add-etcd-registry

wjf222 committed:
fix bug

wjf222 committed:
update grpc version

wjf222 committed:
change jetcd version

kezhenxu94 commented on pom.xml:
```diff
@@ -878,6 +878,31 @@
                 <version>${aws.sdk.version}</version>
             </dependency>
 
+            <dependency>
```
consider importing `grpc-bom` instead

wjf222 committed:
fix SonarCloud Quality Gate failed.

wjf222 committed:
add grpc-pom and fix e2e-workflow error

wjf222 committed:
add grpc-pom and fix e2e-workflow error

wjf222 committed:
add liscense

wjf222 committed:
Merge branch 'dev' into wjf-add-etcd-registry

wjf222 commented:
@kezhenxu94 @caishunfeng @ruanwenjun Hello, my PR has now passed the test, please help to review it, thank you very much

wjf222 commented:
I have followed the suggestions of the previous review and modified it

caishunfeng commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,344 @@
+                .retryDelay(registryProperties.getRetryDelay().toMillis())
+                .retryMaxDelay(registryProperties.getRetryMaxDelay().toMillis())
+                .retryMaxDuration(registryProperties.getRetryMaxDuration());
+        if (!Strings.isNullOrEmpty(registryProperties.getUser()) && (!Strings.isNullOrEmpty(registryProperties.getPassword()))) {
```
```suggestion
        if (StringUtils.isNotEmpty(registryProperties.getUser()) && (StringUtils.isNotEmpty(registryProperties.getPassword()))) {
```

caishunfeng replied:
I only marked one place

caishunfeng commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,344 @@
+        try {
+            List<KeyValue> keyValues = client.getKVClient().get(byteSequence(key)).get().getKvs();
+            return keyValues.iterator().next().getValue().toString(StandardCharsets.UTF_8);
+        } catch (InterruptedException | ExecutionException e) {
```
If throw ExecutionException, should we call `Thread.currentThread().interrupt()`? It's better to handle separately, WDYT?

caishunfeng replied:
I only marked one place

caishunfeng commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,344 @@
+     * @param key the prrefix
+     */
+    @Override
+    public void delete(String key) {
```
It's better to use the more precise method name, such as `deleteByPrefix(String keyPrefix)`

wjf222 replied:
> It's better to use the more precise method name, such as `deleteByPrefix(String keyPrefix)`

This is a function from the Registry interface. It may be a better choice to initiate a new PR to modify the function name.

wjf222 replied:
This is a function from the Registry interface, maybe we can start an Issue to see other people's opinions

wjf222 replied:
I also think deleteByPrefix can better reflect the meaning of the function

caishunfeng commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,344 @@
+
+    /**
+     * delete all keys that contain the prefix
+     * @param key the prrefix
```
```suggestion
     * @param the key prefix
```

caishunfeng commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/test/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistryTest.java:
```diff
@@ -0,0 +1,141 @@
+
+import io.etcd.jetcd.test.EtcdClusterExtension;
+
+public class EtcdRegistryTest {
```
:+1:

wjf222 committed:
handle exceptions separately

wjf222 committed:
Merge remote-tracking branch 'origin/wjf-add-etcd-registry' into wjf-add-etcd-registry

MichaelDeSteven commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdConnectionStateListener.java:
```diff
@@ -0,0 +1,115 @@
+     * Periodically execute thread to get connection status
+     */
+    public void start() {
+        long initialDelay = 500L;
```
why don‘t we design to make it configurable

wjf222 replied:
This parameter is only used inside this function and has nothing to do with etcd clusters.So I didn't make it configurable

MichaelDeSteven commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,338 @@
+import io.etcd.jetcd.watch.WatchEvent;
+
+/**
+ * This is one of the implementation of {@link Registry}, with this implementation, you need to rely on Etcd CLuster to
```
```suggestion
 * This is one of the implementation of {@link Registry}, with this implementation, you need to rely on Etcd cluster to
```

wjf222 committed:
Update dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java

Co-authored-by: SongTao Zhuang <51652084+MichaelDeSteven@users.noreply.github.com>

wjf222 committed:
Merge branch 'dev' into wjf-add-etcd-registry

wjf222 commented:
@kezhenxu94 @caishunfeng @ruanwenjun If there are no other comments, I hope to be able to agree to this PR

ruanwenjun commented on dolphinscheduler-bom/pom.xml:
```diff
@@ -247,6 +250,25 @@
+            <dependency>
+                <groupId>io.etcd</groupId>
+                <artifactId>jetcd-test</artifactId>
+                <version>${jetcd.test.version}</version>
```
if this is only used in UT, please change the scope to test.

ruanwenjun commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-zookeeper/README.md:
```diff
@@ -0,0 +1,27 @@
+
+# How to use
+
+If you want to set the registry center as mysql,you need to set the registry properties in master/worker/api's appplication.yml
```
```suggestion
If you want to set the registry center as zookeeper, you need to set the registry properties in master/worker/api's appplication.yml
```

ruanwenjun commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-zookeeper/README.md:
```diff
@@ -0,0 +1,27 @@
+# Introduction
+
+This module is the etcd registry plugin module, this plugin will use etcd as the registry center.
```
```suggestion
This module is the zookeeper registry plugin module, this plugin will use zookeeper as the registry center.
```

ruanwenjun commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/test/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistryTest.java:
```diff
@@ -0,0 +1,141 @@
+            }
+        }).start();
+        try {
+            preCountDownLatch.await();
```
Please add timeout.

ruanwenjun commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,338 @@
+        try {
+            Long leaseId = threadLocalLockMap.get().get(key);
+            client.getLeaseClient().revoke(leaseId);
+            threadLocalLockMap.get().remove(key);
```
Do we need to check if the `threadLocalLockMap.get()` will null?

wjf222 replied:
If this is null, the remove function will throw an exception

ruanwenjun commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,338 @@
+            Thread.currentThread().interrupt();
+            throw new RegistryException("etcd get lock error", e);
+        } catch (ExecutionException e) {
+            throw new RegistryException("etcd get lock error", e);
```
```suggestion
            throw new RegistryException("etcd get lock error, lockKey: " + key, e);
```

ruanwenjun commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,338 @@
+            Thread.currentThread().interrupt();
+            throw new RegistryException("etcd check key is existed error", e);
+        } catch (ExecutionException e) {
+            throw new RegistryException("etcd check key is existed error", e);
```
```suggestion
            throw new RegistryException("etcd check key is existed error, key: " + key, e);
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdConnectionStateListener.java:
```diff
@@ -0,0 +1,115 @@
+    // monitored client
+    private final Client client;
+    // The state of the last monitor
+    private ConnectionState connectionState;
```
```suggestion
    private volatile ConnectionState connectionState;
```

kezhenxu94 commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,338 @@
+
+    private final Map<String, Watch.Watcher> watcherMap = new ConcurrentHashMap<>();
+
+    private static Long TIME_TO_LIVE_SECONDS = 30L;
```
```suggestion
    private static final long TIME_TO_LIVE_SECONDS = 30L;
```

ruanwenjun commented on dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java:
```diff
@@ -0,0 +1,338 @@
+
+    private static Long TIME_TO_LIVE_SECONDS = 30L;
+    public EtcdRegistry(EtcdRegistryProperties registryProperties) {
+        LOGGER.info("Starting Etcd Registry...");
```
```suggestion
```

wjf222 committed:
update the document and log

wjf222 committed:
await add the timeout

wjf222 commented:
@ruanwenjun I have edited it according to the comments.

wjf222 committed:
Merge branch 'dev' into wjf-add-etcd-registry

wjf222 committed:
implement the function connectUntilTimeout

wjf222 committed:
check style

wjf222 commented:
@kezhenxu94 @caishunfeng @ruanwenjun Please help continue to review. Thank you very much

kezhenxu94 reviewed (commented):
Generally looks good to me, please fix conflict

wjf222 committed:
Merge branch 'dev' into wjf-add-etcd-registry

wjf222 committed:
Update dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java

Co-authored-by: kezhenxu94 <kezhenxu94@apache.org>

wjf222 committed:
Update dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdConnectionStateListener.java

Co-authored-by: kezhenxu94 <kezhenxu94@apache.org>

wjf222 commented:
> Generally looks good to me, please fix conflict

@kezhenxu94 Thanks, I have fixed the conflict

caishunfeng commented:
@wjf222 Please fix the conflicts.

wjf222 committed:
Merge branch 'dev' into wjf-add-etcd-registry

wjf222 commented:
> Thanks, I have fixed the conflict

@caishunfeng I have fixed the conflict. Test bugs are due to new 3rd party packages introduced in other commits

wjf222 commented:
@caishunfeng Please pass the review,  I have fixed the conflict. Thank you.

wjf222 commented:
@kezhenxu94 @caishunfeng @ruanwenjun Due to previous conflicts, modifications have been made, and I hope to pass this review again. Thank you very much.

ruanwenjun commented:
@wjf222 Please resolve the dependency check failure.

wjf222 committed:
Merge branch 'apache:dev' into wjf-add-etcd-registry

wjf222 committed:
fix checkstyle

wjf222 commented:
> @wjf222 Please resolve the dependency check failure.

@ruanwenjun  I have fixed the conflict.  Thank you.

wjf222 commented:
@kezhenxu94 @caishunfeng @ruanwenjun I have passed the test.Please apprve the review.Thank you.

wjf222 commented:
@kezhenxu94 @caishunfeng @ruanwenjun I have passed the test.Please apprve the review.Thank you.

ruanwenjun reviewed (approved):
LGTM, I will merged this to dev branch, and this feature will be released at 3.1.0, thanks for your contribution. @wjf222
